### PR TITLE
fix: retry specific internal errors

### PR DESCRIPTION
**Pull request by olavloite**
Some specific internal errors should be retrid. Instead of adding INTERNAL
as a standard retryable error code, we use an interceptor to catch and
translate those specific errors.

See also b/375684610

### CODE REVIEW

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerErrorInterceptor.java [hunk 1]
```diff
@@ -69,6 +71,12 @@ public void start(Listener<RespT> responseListener, Metadata headers) {
               @Override
               public void onClose(Status status, Metadata trailers) {
                 try {
+                  // Translate INTERNAL errors that should be retried to a retryable error code.
+                  if (status.getCode() == Code.INTERNAL
```
Comment (sakthivelmanii) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerErrorInterceptor.java:75]: I can see the same check in `isRetryableInternalError` as well. Can we remove it from here?

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/RetryableInternalErrorTest.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      }
+      assertEquals(
+          Long.valueOf(1L),
+          client
+              .readWriteTransaction()
+              .run(transaction -> transaction.executeUpdate(INSERT_STATEMENT)));
+
+      // Verify that all the RPCs were retried.
+      assertEquals(2, mockSpanner.countRequestsOfType(BatchCreateSessionsRequest.class));
+      assertEquals(4, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
```
Comment (sakthivelmanii) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/RetryableInternalErrorTest.java:77]: Do you know why `ExecuteSqlRequest` is sent 4 times? We have setup exception only once for ExecuteSql right?

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerErrorInterceptor.java (hunk 1 above)
Comment (olavloite) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/SpannerErrorInterceptor.java]: yeah, good point. I initially had this set up in a way where only the error message was checked in the RetryableInternalError class, and forgot to remove this after that refactor. Removed now.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/RetryableInternalErrorTest.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+              Status.INTERNAL
+                  .withDescription("Authentication backend internal server error. Please retry.")
+                  .asRuntimeException()));
+      mockSpanner.setExecuteSqlExecutionTime(
+          SimulatedExecutionTime.ofException(
+              Status.INTERNAL
+                  .withDescription("Authentication backend internal server error. Please retry.")
+                  .asRuntimeException()));
+
+      try (ResultSet resultSet = client.singleUse().executeQuery(SELECT1_STATEMENT)) {
```
[the remark this answers is not in the document]
Comment (olavloite) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/RetryableInternalErrorTest.java:62]: I've re-organized the test a bit to show that both `ExecuteStreamingSql` and `ExecuteSql` are retried.

**Review by sakthivelmanii — APPROVED** [review]

**Pull request merged**